tools/src/scie_pants/pants_version.py

```diff
@@ -162,5 +182,120 @@ def configure_version():
         pants_config.write_text(tomlkit.dumps(config))
 
     return configure_version, determine_tag_version(
-        ptex, pants_version, find_links_dir, github_api_bearer_token
+        ptex, pants_version, find_links_dir, github_api_bearer_token, bootstrap_urls_path
     )
+
+
+PYTHON_IDS = {
+    # N.B.: These values must match the lift TOML interpreter ids.
+    # Important: all pythons used in pants_python_versions.json must be represented in this list.
+    "cp38": "cpython38",
+    "cp39": "cpython39",
+    "cp310": "cpython310",
+    "cp311": "cpython311",
+}
+
+
+def determine_pex_url_and_python_id(
+    ptex: Ptex,
+    version: Version,
+    bootstrap_urls_path: str | None,
+) -> tuple[str, str]:
+    uname = os.uname()
+    platform = f"{uname.sysname.lower()}_{uname.machine.lower()}"
+    res = get_pex_url_and_python_id(ptex, version, platform, bootstrap_urls_path)
+    if not res:
+        bootstrap_urls_info = (
+            (
+                f"It may also be a missing Pants PEX entry in the PANTS_BOOTSTRAP_URLS file {bootstrap_urls_path} for "
+                f"a name like `pants.{version}-cpNN-{platform}.pex`.\n\n"
+            )
+            if bootstrap_urls_path
+            else ""
+        )
+        fatal(
+            f"Unknown Pants release: {version}.\n\n"
+            "If this is unexpected (you are using a known good Pants version), try upgrading scie-pants first.\n"
+            f"It may also be that the platform {platform} isn't supported for this version of Pants.\n"
+            + bootstrap_urls_info
+            + "To get help, please visit: https://www.pantsbuild.org/community/getting-help"
+        )
+    pex_url, python = res
+    if python not in PYTHON_IDS:
+        # Should not happen... but if we mess up, this is a nicer error message rather than blowing up.
+        fatal(f"This version of scie-pants does not support {python!r}.")
+    return pex_url, PYTHON_IDS[python]
+
+
+def get_pex_url_and_python_id(
+    ptex: Ptex,
+    version: Version,
+    platform: str,
+    bootstrap_urls_path: str | None,
+) -> tuple[str, str] | None:
+    if bootstrap_urls_path:
+        bootstrap_urls = json.loads(Path(bootstrap_urls_path).read_text())
+        ptex_urls = bootstrap_urls.get("ptex")
+        if ptex_urls is None:
+            raise ValueError(
+                f"Missing 'ptex' key in PANTS_BOOTSTRAP_URLS file: {bootstrap_urls_path}"
+            )
+    else:
+        ptex_urls = None
```

Comment:
What do you think about putting all of the bootstrap URLs handling here, and then the rest of this function doesn't have to think about it? 

In particular, if someone specifies this file, the intent is "do all downloads from these URLs". If a file is missing, we should give an error about that, but currently this new code will fallback to the URL-probing (i.e. doing external requests) which is a bit surprising.

I think we can also "trust" those URLs and don't need to validate them, particularly, which eliminates the `file://` URL handling that was previously required.

So, maybe something like:

```suggestion
    if bootstrap_urls_path:
         bootstrap_urls = json.loads(Path(bootstrap_urls_path).read_text())
         ptex_urls = bootstrap_urls.get("ptex")
         if ptex_urls is None:
             raise ValueError(
                 f"Missing 'ptex' key in PANTS_BOOTSTRAP_URLS file: {bootstrap_urls_path}"
             )

         # do all the validation that was previously in `install_pants.py`
         return ...
```

And then remove the `ptex_urls` arg to `get_download_url`.

Reply:
> [...] but currently this new code will fallback to the URL-probing (i.e. doing external requests) which is a bit surprising.

I don't think this is correct. What it does is, it checks that the provided URL actually works before accepting it. If there is no URL provided for a particular platform/python we get `None` and proceed with trying all known Pythons etc, but `get_download_url` will never hit the network if you use the bootstrap urls file (for any other url than provided by that file).